### ISSUE TITLE
[Feature]: customize LSP source: hide word under cursor from bar

### ISSUE THREAD

**Issue by disrupted** [Bekaboo/dropbar.nvim#202]
### Problem

First of all, thanks for this amazing plugin! I am currently in the process of testing it out and customizing it for my workflow. Ideally, I want the breadcrumbs to quickly tell me in which file + class + method I am without too much noise. Naturally, I disabled most types for the Treesitter source except the relevant ones.

Now I understand that Treesitter falls short in some places, so I am sticking to the default where LSP takes priority over Treesitter. Whenever LSP document symbols are used I notice that it also adds the symbol under the cursor to the end of the breadcrumbs bar, e.g. the parameter of a function. Is there a way to disable this? For my personal workflow it doesn't add much value, so I'd rather hide it.

### Expected behavior

A setting to disable the cursor word or customize the LSP source

**Comment by disrupted** [comment]
here is a quick mockup of what I have in mind

```patch
diff --git a/lua/dropbar/configs.lua b/lua/dropbar/configs.lua
index 28647c9..d049b48 100644
--- a/lua/dropbar/configs.lua
+++ b/lua/dropbar/configs.lua
@@ -664,6 +664,16 @@ M.opts = {
       },
     },
     lsp = {
+      ignore_symbols = {
+        'Variable',
+        'Constant',
+        'String',
+        'Number',
+        'Boolean',
+        'Keyword',
+        'Null',
+        'EnumMember',
+      },
       request = {
         -- Times to retry a request before giving up
         ttl_init = 60,
diff --git a/lua/dropbar/sources/lsp.lua b/lua/dropbar/sources/lsp.lua
index 14c2a63..4c133d8 100644
--- a/lua/dropbar/sources/lsp.lua
+++ b/lua/dropbar/sources/lsp.lua
@@ -83,6 +83,17 @@ local symbol_kind_names = setmetatable({
 })
 -- stylua: ignore end
 
+local function reverse_table(tbl)
+  local reversed = {}
+  for i = #tbl, 1, -1 do
+    table.insert(reversed, tbl[i])
+  end
+  return reversed
+end
+
+-- Map symbol kind to symbol number
+local symbol_names_kind = reverse_table(symbol_kind_names)
+
 ---Return type of the symbol table
 ---@param symbols lsp_symbol_t[] symbol table
 ---@return string? symbol type
@@ -210,10 +221,17 @@ local function convert_document_symbol_list(
   -- is preferred
   for idx, symbol in vim.iter(lsp_symbols):enumerate():rev() do
     if cursor_in_range(cursor, symbol.range) then
-      table.insert(
-        dropbar_symbols,
-        convert_document_symbol(symbol, buf, win, lsp_symbols, idx)
-      )
+      if
+        not vim.tbl_contains(
+          configs.opts.sources.lsp.ignore_symbols,
+          symbol_names_kind[symbol.kind]
+        )
+      then
+        table.insert(
+          dropbar_symbols,
+          convert_document_symbol(symbol, buf, win, lsp_symbols, idx)
+        )
+      end
       if symbol.children then
         convert_document_symbol_list(
           symbol.children,
```

**Comment by disrupted** [comment]
here are some examples

<img width="832" alt="screen 2024-12-18 at 09 52 44" src="https://github.com/user-attachments/assets/a2e759d4-3866-470d-a7bc-b6be55e4f615" />
<img width="617" alt="screen 2024-12-18 at 09 53 47" src="https://github.com/user-attachments/assets/cde5c44e-42c3-4b2e-a6ad-caff4fa8d69c" />
<img width="760" alt="screen 2024-12-18 at 09 53 26" src="https://github.com/user-attachments/assets/6cd0c8cd-fa67-4a34-9561-233eff9d89e6" />

basically I want to use dropbar only for displaying in which class + method I am right now

**Comment by Bekaboo** [comment]
I get you. How about configuring symbols that should be included instead of those need to be excludeed/ignored to align with the treesitter `valid_types` option?

**Comment by disrupted** [comment]
sounds good to me! I wasn't sure but agree that it should be aligned with the other sources. Do you want me to open a PR for it?

**Comment by Bekaboo** [comment]
> Do you want me to open a PR for it?

Yes please.

**Comment by disrupted** [comment]
despite my changes I still haven't figured out to get rid of all the cases I showcased above.

for example with
```lua
valid_types = {
  'Class',
  'Method',
  'Function',
}
```

it still shows a constant here
<img width="462" alt="screen 2024-12-18 at 23 20 51" src="https://github.com/user-attachments/assets/cca87c8f-c1d0-4b8e-9efb-13154ed025fa" />

**Comment by disrupted** [comment]
I made some further tweaks to make it work the way I envisioned. Now during window creation, dropbar will initially register the Treesitter source if no LS exists that supports document symbols. But during evaluation the LS might not have attached yet, so I created an autocmd that will overwrite the sources and refresh dropbar whenever the LS supporting document symbols attaches.

```lua
-- relevant parts from my lazy config for dropbar
---@type dropbar_configs_t
opts = {
    bar = {
        ---@type dropbar_source_t[]|fun(buf: integer, win: integer): dropbar_source_t[]
        sources = function(buf, _)
            local sources = require 'dropbar.sources'
            if vim.bo[buf].ft == 'markdown' then
                return {
                    sources.path,
                    sources.markdown,
                }
            end
            if vim.bo[buf].buftype == 'terminal' then
                return {
                    sources.terminal,
                }
            end

            local clients = vim.lsp.get_clients {
                bufnr = buf,
                method = require('vim.lsp.protocol').Methods.textDocument_documentSymbol,
            }
            if not vim.tbl_isempty(clients) then
                return {
                    sources.path,
                    sources.lsp,
                }
            else
                return {
                    sources.path,
                    sources.treesitter,
                }
            end
        end,
    },
    sources = {
        lsp = {
            valid_symbols = {
                'Class',
                'Method',
                'Function',
            },
        },
    },
},
config = function(_, opts)
    require('dropbar').setup(opts)

    vim.api.nvim_create_autocmd('LspAttach', {
        desc = 'Register LSP source and disable Treesitter source when an LS that supports documentSymbol attaches.',
        group = vim.api.nvim_create_augroup(
            'DropBarLspAttachRefreshSources',
            { clear = true }
        ),
        callback = function(args)
            local client = vim.lsp.get_client_by_id(args.data.client_id)
            if
                client
                and client.supports_method 'textDocument/documentSymbol'
            then
                local sources = require 'dropbar.sources'
                local utils = require 'dropbar.utils'
                for win in
                    pairs(
                        _G.dropbar.bars[vim.api.nvim_get_current_buf()]
                    )
                do
                    _G.dropbar.bars[vim.api.nvim_get_current_buf()][win].sources =
                        {
                            sources.path,
                            sources.lsp,
                        }
                    utils.bar.exec('update', { win = win })
                end
            end
        end,
    })
end
```
